rflx/specification/integration.py

```diff
@@ -0,0 +1,137 @@
+
+    def add_integration_file(self, filename: Path, file: object) -> None:
+        try:
+            f = IntegrationFile.parse_obj(file)
+            self._files[filename] = f
+        except ValidationError as e:
+            fail(f"{filename}: {str(e)}", Subsystem.PARSER, Severity.ERROR, None)
+
+    @staticmethod
+    def __validate_globals(
+        filename: Path, integration: SessionIntegration, session: Session
+    ) -> None:
+        if integration.buffer_size.global_ is None:
+            return
+        for var_name in integration.buffer_size.global_.keys():
+            if var_name not in [x.parts[-1] for x in session.declarations.keys()]:
+                fail(
+                    (
+                        f'{filename}: unknown variable "{var_name}" in global state '
+                        f'of session "{session.identifier.parts[-1]}"'
+                    ),
+                    Subsystem.PARSER,
+                    Severity.ERROR,
+                    None,
+                )
+
+    @staticmethod
+    def __validate_states(
+        filename: Path, integration: SessionIntegration, session: Session
+    ) -> None:
+        if integration.buffer_size.local_ is None:
+            return
+        for state_name, state_entry in integration.buffer_size.local_.items():
+            state = None
+            for s in session.states:
+                if s.identifier.parts[-1] == state_name:
+                    state = s
+            if state is None:
+                fail(
+                    (
+                        f'{filename}: unknown state "{state_name}" in session '
+                        f'"{session.identifier.parts[-1]}"'
+                    ),
+                    Subsystem.PARSER,
+                    Severity.ERROR,
+                    None,
+                )
+            for var_name in state_entry:
+                if var_name not in [x.parts[-1] for x in state.declarations.keys()]:
+                    fail(
+                        (
+                            f'{filename}: unknown variable "{var_name}" in state '
+                            f'"{state_name}" of session "{session.identifier.parts[-1]}"'
+                        ),
+                        Subsystem.PARSER,
+                        Severity.ERROR,
+                        None,
+                    )
+
+    def validate(self, model: Model) -> None:
+        for filename, file in self._files.items():
+            for session_name, integration in file.session.items():
+                session = None
+                for s in model.sessions:
+                    if s.identifier.parts[-1] == session_name:
+                        session = s
+                if (
+                    session is None
+                    or session.location is None
+                    or session.location.source is None
+                    or session.location.source.with_suffix(".rfi") != filename
+                ):
+                    fail(
```

Comment:
`model.sessions` can contain multiple sessions with the same name, but different packages. For finding the corresponding session, the current code just checks the name (`.parts[-1]`), but not the package. This will lead to false positives.

Reply:
You are correct. The code currently checks validity of the .rfi file after all sessions and all .rfi files have been parsed. Putting things together at the end isn't easy. I will try to come up with something.

Reply:
I'm not sure if we are talking about the same thing. The problem is that checking `s.identifier.parts[-1] == session_name` is not sufficient to find the corresponding session object. `s.package` must be also checked against `filename`.

To ensure the correct behavior, please add a test with two specifications (e.g., `a.rflx` and `b.rflx`) and two integration files (e.g., `a.rfi` and `b.rfi`). In both specifications, a session with the same name (e.g., `Session`) should be defined. In both integration files, buffer sizes of the session should be defined. It shall be tested that the buffer sizes are associated with the right session and that there is no false positive error.

Reply:
Yes I understood you correctly. I have added such a test and fixed the code in 9da6e9cbe2ebd35dff278dcd2ab3a326aa2f1fd8.
Note that I'm still not looking at package names, only file names. But my understanding is that this is correct as there is only one package per file.

Reply:
The fixed code seems to have the intended behavior.